crates/cards/src/validate.rs

```diff
@@ -0,0 +1,146 @@
+use std::{fmt, ops::Deref, str::FromStr};
+
+use masking::{PeekInterface, Strategy, StrongSecret, WithType};
+use serde::{Deserialize, Deserializer, Serialize};
+use thiserror::Error;
+
+#[derive(Debug, Deserialize, Serialize, Error)]
+#[error("not a valid credit card number")]
+pub struct CCValError;
+
+impl From<core::convert::Infallible> for CCValError {
+    fn from(_: core::convert::Infallible) -> Self {
+        Self
+    }
+}
+
+/// Card number
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize)]
+pub struct CardNumber(StrongSecret<String, CardNumberStrategy>);
+
+impl FromStr for CardNumber {
+    type Err = CCValError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match luhn::valid(s) {
+            true => Ok(Self(StrongSecret::from_str(s)?)),
+            false => Err(CCValError),
+        }
+    }
+}
+
+impl TryFrom<String> for CardNumber {
+    type Error = CCValError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::from_str(&value)
+    }
+}
+
+impl Deref for CardNumber {
+    type Target = StrongSecret<String, CardNumberStrategy>;
+
+    fn deref(&self) -> &StrongSecret<String, CardNumberStrategy> {
+        &self.0
+    }
+}
+
+impl<'de> Deserialize<'de> for CardNumber {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let s = String::deserialize(d)?;
+        Self::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+impl CardNumber {
+    pub fn remove_whitespace(self) -> Self {
+        let cc_no_whitespace: String = self.peek().split_whitespace().collect();
+        let secret = StrongSecret::new(cc_no_whitespace);
+        Self(secret)
+    }
+}
```

Comment:
@Narayanbhat166 Can we remove all whitespaces when constructing a `CardNumber` itself, or are there any connectors which require us to send card numbers with spaces?

Reply:
It's not usually the case that connectors accept card numbers with spaces only, without spaces also it should work as expected. But if there arises any case where the connector requires the card number to be sent with spaces only then we can have the conversion to add whitespaces for that connector only. To be consistent, I would say better to remove all the whitespaces. Currently with the connectors that I have come across, they accept card numbers without spaces. @jagan-jaya can share his opinion on this.

Reply:
I did discuss this with @jagan-jaya, we can go ahead and remove spaces when constructing the `CardNumber` type. @phillyphil91 Please make that change, and remove the `remove_whitespace()` method you've added here.

Reply:
So basically remove all whitespaces in the from_str implementation of CardNumber so that at the end we will always have a CC number without whitespaces? (If the number is valid in general of course)